lgc/interface/lgc/ElfLinker.h

```diff
@@ -0,0 +1,84 @@
+/**
+ ***********************************************************************************************************************
+ * @file  ElfLinker.h
+ * @brief LLPC header file: LGC interface for linking unlinked half-pipeline ELFs into pipeline ELF
+ ***********************************************************************************************************************
+ */
+#pragma once
+
+#include "lgc/CommonDefs.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/StringRef.h"
+
+namespace lgc {
+
+// =====================================================================================================================
+// The public API of the LGC interface for ELF linking.
+// The ElfLinker object is created by calling Pipeline::getElfLinker(). The ElfLinker internally refers back to
+// its Pipeline, and thus uses pipeline state for glue code generation, adding to PAL metadata, and resolving
+// relocs.
+class ElfLinker {
+public:
+  // Get information on the glue code that will be needed for the link. It is an implementation detail how many
+  // chunks of glue there might be and what they are for, but, for information, they will be some subset of:
+  // - A CS prolog
+  // - A VS prolog ("fetch shader")
+  // - A vertex-processing epilog ("parameter export shader")
+  // - An FS epilog ("color export shader")
+  //
+  // Returns an array (possibly 0 length) with an entry for each chunk of glue code, where an entry
+  // is a StringRef that the client can hash for its cache lookup. If it gets a cache hit, it should provide the
+  // found blob to ElfLinker::addGlue. If it does not get a cache hit, the client can call ElfLinker::compileGlue to
+  // retrieve the compiled glue code to store in the cache.
+  virtual llvm::ArrayRef<llvm::StringRef> getGlueInfo() = 0;
+
+  // Add a blob for a particular chunk of glue code, typically retrieved from a cache. The blob is not copied,
+  // and remains in use until the first of the link completing or the ElfLinker's parent Pipeline being destroyed.
+  //
+  // @param glueIndex : Index into the array that was returned by getGlueInfo()
+  // @param blob : Blob for the glue code
+  virtual void addGlue(unsigned glueIndex, llvm::StringRef blob) = 0;
+
+  // Compile a particular chunk of glue code and retrieve its blob. The returned blob remains valid until the first
+  // of calling link() or the ElfLinker's parent Pipeline being destroyed. It is optional to call this; any chunk
+  // of glue code that has not had one of addGlue() or compileGlue() done by the time link() is called will be
+  // internally compiled. The client only needs to call this if it wants to cache the glue code's blob.
+  //
+  // @param glueIndex : Index into the array that was returned by getGlueInfo()
+  // @return : The blob. A zero-length blob indicates that a recoverable error occurred, and link() will also return
+  //           and empty ELF blob.
+  virtual llvm::StringRef compileGlue(unsigned glueIndex) = 0;
```

Comment:
StringRef isn't really the right type for this, is it?

Reply:
It seems to be the done thing. It is common to pass around a MemoryBufferRef, which is a StringRef with an attached name for error reporting etc. So the use of StringRef to represent a read-only binary blob whose memory you don't own does seem right.

Reply:
Ah I see, you're right.